dash-pipeline/SAI/src/p4meta.cpp

```diff
@@ -0,0 +1,361 @@
+    {
+        if (meta_key->ip_is_v6_field_id) {
+            auto mf = matchActionEntry->add_match();
+            mf->set_field_id(meta_key->ip_is_v6_field_id);
+            set_attr_ipaddr_family_to_p4(attr->value, mf->mutable_exact());
+        }
+
+        auto mf = matchActionEntry->add_match();
+        mf->set_field_id(meta_key->id);
+        if (meta_key->match_type == "ternary" && string_has_suffix(meta_key->name, "_MASK")) {
+            set_attr_value_mask_to_p4_ternary(meta_key->field, meta_key->bitwidth, attr->value,
+                     mf->mutable_ternary());
+        }
+        else {
+            set_attr_value_to_p4_match(*meta_key, attr->value, mf);
+        }
+    }
+
+    void  set_attr_to_p4_action(
+            _In_ const P4MetaActionParam *meta_param,
+            _In_ const sai_attribute_t *attr,
+            _Out_ p4::v1::Action *action)
+    {
+        if (meta_param->ip_is_v6_field_id) {
+            auto param = action->add_params();
+            param->set_param_id(meta_param->ip_is_v6_field_id);
+            set_attr_ipaddr_family_to_p4(attr->value, param);
+        }
+
+        auto param = action->add_params();
+        param->set_param_id(meta_param->id);
+        set_attr_value_to_p4(meta_param->field, meta_param->bitwidth, attr->value, param);
+    }
+
+    std::pair<p4::v1::FieldMatch*, p4::v1::FieldMatch*> get_match_pair_from_p4_table_entry(
+            _In_ const P4MetaKey *meta_key,
+            _In_ std::shared_ptr<p4::v1::TableEntry> matchActionEntry)
+    {
+        std::pair<p4::v1::FieldMatch*, p4::v1::FieldMatch*> pair_key = {nullptr, nullptr};
+
+        for (int i = 0; i < matchActionEntry->match_size(); i++) {
+            auto mf = matchActionEntry->mutable_match(i);
+            if (mf->field_id() == meta_key->id) {
+                pair_key.first = mf;
+            }
+            else if (mf->field_id() == meta_key->ip_is_v6_field_id) {
+                pair_key.second = mf;
+            }
+        }
+
+        return pair_key;
+    }
+
+    std::pair<p4::v1::Action_Param*, p4::v1::Action_Param*> get_action_param_pair_from_p4_table_entry(
+            _In_ const P4MetaActionParam *meta_param,
+            _In_ std::shared_ptr<p4::v1::TableEntry> matchActionEntry)
+    {
+        auto action = matchActionEntry->mutable_action()->mutable_action();
+        std::pair<p4::v1::Action_Param*, p4::v1::Action_Param*> pair_param = {nullptr, nullptr};
+
+        for (int i = 0; i < action->params_size(); i++) {
+            auto param = action->mutable_params(i);
+            if (param->param_id() == meta_param->id) {
+                pair_param.first = param;
+            }
+            else if (param->param_id() == meta_param->ip_is_v6_field_id) {
+                pair_param.second = param;
+            }
+        }
+
+        return pair_param;
+    }
+
+    bool string_has_suffix(const std::string &str, const std::string &suffix)
```

Comment:
string_ends_with

Reply:
fix it at https://github.com/sonic-net/DASH/pull/653